### PR TITLE
User update form

### LINKED ISSUE #29: User update form

### PR DESCRIPTION
**Pull request by Katrin-kudryash**
## PR includes

-   [x] Feature 

## Related issues

Resolve #29

### CODE REVIEW

**Comment by 9teen90nine** [comment]
image

File: src/components/users/UserProfile.tsx [hunk 1]
```diff
@@ -114,6 +119,7 @@ export const UserProfile = () => {
     const { userId } = router.query;
     const userQuery = useUser(String(userId));
     const user = userQuery.data;
+    const [openUserUpdateForm, setUserUpdateForm] = useState(false);
```
Comment (nyet-ty) [inline src/components/users/UserProfile.tsx:122]: maybe setOpenUpdateUserForm?

File: src/components/users/form/UserUpdateForm.tsx [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        defaultValues: {
+            fullName: user.fullName,
+            email: user.email,
+            supervisorId: user.supervisor?.userId,
+            github: user.github,
+            telegram: user.telegram,
+        },
+        resolver: zodResolver(userUpdateSchema),
+    });
+    const userUpdate = trpc.user.update.useMutation();
```
Comment (nyet-ty) [inline src/components/users/form/UserUpdateForm.tsx:36]: maybe better to move it and updateUser to user-api-hook?

**Review by nyet-ty — APPROVED** [review]

File: src/components/users/form/UserUpdateForm.tsx (hunk 2 above)
Comment (9teen90nine) [reply src/components/users/form/UserUpdateForm.tsx:36]: No, we'll get rid of that file completely as it currently used only on the back-end

**Review by awinogradov — CHANGES_REQUESTED** [review]
Just to not miss...

File: src/api-client/users/update-user-dto.d.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+export interface UpdateUserDto {
+    supervisorId?: string;
+    github?: string;
+    telegram?: string;
```
Comment (9teen90nine) [inline src/api-client/users/update-user-dto.d.ts:4]: let's delete unused fields here

File: src/trpc/router/user-router.ts [hunk 4]
```diff
@@ -45,4 +46,7 @@ export const userRouter = router({
         accessCheck(userAccess.isBonusHistoryViewable(ctx.session.user, input));
         return userMethods.getBonusPointsHistory(input);
     }),
+    edit: protectedProcedure.input(editUserSchema).mutation(({ input }) => {
```
Comment (tsumo) [inline src/trpc/router/user-router.ts:49]: Should have access check for isEditable

File: src/components/users/UserSupervisorEditable.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+`;
+
+const StyledInputsWrapper = styled.div`
+    display: grid;
+    grid-template-columns: 1fr 1fr max-content;
+`;
+
+const StyledComboBox = styled(ComboBox)`
+    width: fit-content;
+    padding-left: 2px;
```
[the remark this answers is not in the document]
Comment (Katrin-kudryash) [reply src/components/users/UserSupervisorEditable.tsx]: no. I deleted

File: src/components/users/UserSupervisorEditable.tsx [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+`;
+
+interface UserSupervisorEditableProps {
+    user: User;
```
[the remark this answers is not in the document]
Comment (Katrin-kudryash) [reply src/components/users/UserSupervisorEditable.tsx]: Show input with a list of users

File: src/modules/user.schemas.ts [hunk 7]
```diff
@@ -33,3 +33,11 @@ export const getUserListSchema = z.object({
         .optional(),
 });
 export type GetUserList = z.infer<typeof getUserListSchema>;
+
+export const editUserSchema = z.object({
+    id: z.string(),
+    name: z.string().nullish().optional(),
+    supervisorId: z.string().nullish().optional(),
```
Comment (tsumo) [inline src/modules/user.schemas.ts:40]: You don't need `optional` if you already have `nullish`

**Review by tsumo — CHANGES_REQUESTED** [review]

**Comment by tsumo** [comment]
Rebase, please

File: src/components/users/UserSupervisorEditable.tsx [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+import { gray10, gray7 } from '@taskany/colors';
+import styled from 'styled-components';
+import { useState } from 'react';
+import { User } from 'prisma/prisma-client';
+
+import { trpc } from '../../trpc/trpcClient';
+
+import { tr } from './users.i18n';
+
+const StyledInput = styled(Input)`
```
Comment (awinogradov) [inline src/components/users/UserSupervisorEditable.tsx:11]: Looks like this is the full override. I don't see reasons to do that here.

File: src/components/users/UserSupervisorEditable.tsx [hunk 9]
```diff
@@ -0,0 +1,73 @@
+import { ComboBox, Input, MenuItem } from '@taskany/bricks';
+import { gray10, gray7 } from '@taskany/colors';
+import styled from 'styled-components';
+import { useState } from 'react';
+import { User } from 'prisma/prisma-client';
```
Comment (awinogradov) [inline src/components/users/UserSupervisorEditable.tsx:5]: We have issue about ban to do import of prisma interfaces in client code.

File: src/components/users/form/UserUpdateForm.tsx [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    FormInput,
+    FormTitle,
+    ModalHeader,
+    ModalCross,
+    Text,
+    nullable,
+} from '@taskany/bricks';
+import { zodResolver } from '@hookform/resolvers/zod';
+import styled from 'styled-components';
+import { User } from 'prisma/prisma-client';
```
Comment (awinogradov) [inline src/components/users/form/UserUpdateForm.tsx:17]: Banned import

File: src/components/users/form/UserUpdateForm.tsx [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+import { gray8 } from '@taskany/colors';
+
+import { UserSupervisorEditable } from '../UserSupervisorEditable';
+import { EditUser, editUserSchema } from '../../../modules/user.schemas';
+import { useUserMutations } from '../../../modules/user.hooks';
+import { UserMeta, UserSupervisor } from '../../../modules/user.types';
+
+import { tr } from './form.i18n';
+
+type UserDataFormProps = {
```
Comment (awinogradov) [inline src/components/users/form/UserUpdateForm.tsx:27]: Must be interface

File: src/components/users/form/UserUpdateForm.tsx [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+const NoWrap = styled.div`
+    white-space: nowrap;
+`;
+
+const StyledSupervisorField = styled.div`
+    display: inline-flex;
+`;
+
+const StyledTextSupervisor = styled(Text)`
+    padding-left: 15px;
```
Comment (awinogradov) [inline src/components/users/form/UserUpdateForm.tsx:41]: What the numbers? We have gaps in bricks — use them instead.

File: src/components/users/form/UserUpdateForm.tsx [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+                    <NoWrap>
+                        <FormInput
+                            label={tr('Full Name')}
+                            {...register('name')}
+                            flat="bottom"
+                            brick="right"
+                            autoComplete="off"
+                        />
+                        <StyledSupervisorField>
+                            <StyledTextSupervisor weight={'bold'} color={gray8}>
```
Comment (awinogradov) [inline src/components/users/form/UserUpdateForm.tsx:87]: We have lint rules to not write `weight={'bold'}`. It must be `weight="bold"`

**Review by awinogradov — CHANGES_REQUESTED** [review]

File: src/components/users/UserSupervisorEditable.tsx (hunk 9 above)
Comment (tsumo) [reply src/components/users/UserSupervisorEditable.tsx:5]: Enabling the rule will trigger large scale refactoring, which will be outside the scope of this issue. It should be done separately
https://github.com/taskany-inc/crew/issues/288

**Review by awinogradov — APPROVED** [review]

**Review by 9teen90nine — APPROVED** [review]

**Review by tsumo — APPROVED** [review]

**Pull request merged**